rosidl_typesupport_introspection_tests/test/test_typesupport_c.cpp

```diff
@@ -0,0 +1,2215 @@
+  EXPECT_EQ(members[10].is_upper_bound_, true);
+  EXPECT_EQ(
+    members[10].offset_,
+    offsetof(MESSAGE_STRUCT(test_msgs, msg, BoundedSequences), uint32_values));
+  EXPECT_EQ(members[10].default_value_, nullptr);
+  EXPECT_EQ(members[10].size_function, nullptr);
+  EXPECT_EQ(members[10].get_const_function, nullptr);
+  EXPECT_EQ(members[10].get_function, nullptr);
+  EXPECT_EQ(members[10].resize_function, nullptr);
+
+  EXPECT_STREQ(members[11].name_, "int64_values");
+  EXPECT_EQ(members[11].type_id_, rosidl_typesupport_introspection_c__ROS_TYPE_INT64);
+  EXPECT_EQ(members[11].string_upper_bound_, 0u);
+  EXPECT_EQ(members[11].members_, nullptr);
+  EXPECT_EQ(members[11].is_array_, true);
+  EXPECT_EQ(members[11].array_size_, 3u);
+  EXPECT_EQ(members[11].is_upper_bound_, true);
+  EXPECT_EQ(
+    members[11].offset_,
+    offsetof(MESSAGE_STRUCT(test_msgs, msg, BoundedSequences), int64_values));
+  EXPECT_EQ(members[11].default_value_, nullptr);
+  EXPECT_EQ(members[11].size_function, nullptr);
+  EXPECT_EQ(members[11].get_const_function, nullptr);
+  EXPECT_EQ(members[11].get_function, nullptr);
+  EXPECT_EQ(members[11].resize_function, nullptr);
+
+  EXPECT_STREQ(members[12].name_, "uint64_values");
+  EXPECT_EQ(members[12].type_id_, rosidl_typesupport_introspection_c__ROS_TYPE_UINT64);
+  EXPECT_EQ(members[12].string_upper_bound_, 0u);
+  EXPECT_EQ(members[12].members_, nullptr);
+  EXPECT_EQ(members[12].is_array_, true);
+  EXPECT_EQ(members[12].array_size_, 3u);
+  EXPECT_EQ(members[12].is_upper_bound_, true);
+  EXPECT_EQ(
+    members[12].offset_,
+    offsetof(MESSAGE_STRUCT(test_msgs, msg, BoundedSequences), uint64_values));
+  EXPECT_EQ(members[12].default_value_, nullptr);
+  EXPECT_EQ(members[12].size_function, nullptr);
+  EXPECT_EQ(members[12].get_const_function, nullptr);
+  EXPECT_EQ(members[12].get_function, nullptr);
+  EXPECT_EQ(members[12].resize_function, nullptr);
+
+  EXPECT_STREQ(members[13].name_, "string_values");
+  EXPECT_EQ(members[13].type_id_, rosidl_typesupport_introspection_c__ROS_TYPE_STRING);
+  EXPECT_EQ(members[13].string_upper_bound_, 0u);
+  EXPECT_EQ(members[13].members_, nullptr);
+  EXPECT_EQ(members[13].is_array_, true);
+  EXPECT_EQ(members[13].array_size_, 3u);
+  EXPECT_EQ(members[13].is_upper_bound_, true);
+  EXPECT_EQ(
+    members[13].offset_,
+    offsetof(MESSAGE_STRUCT(test_msgs, msg, BoundedSequences), string_values));
+  EXPECT_EQ(members[13].default_value_, nullptr);
+  EXPECT_EQ(members[13].size_function, nullptr);
+  EXPECT_EQ(members[13].get_const_function, nullptr);
+  EXPECT_EQ(members[13].get_function, nullptr);
+  EXPECT_EQ(members[13].resize_function, nullptr);
+
+  EXPECT_STREQ(members[16].name_, "defaults_values");
+  EXPECT_EQ(members[16].type_id_, rosidl_typesupport_introspection_c__ROS_TYPE_MESSAGE);
+  EXPECT_EQ(members[16].string_upper_bound_, 0u);
+  // TODO(gbiggs): Test the contained members are correct to verify this field was initialised
+  // correctly
```

Comment:
@gbiggs hmm, perhaps it makes sense to have `*AreCorrect` test cases validate members only, and then add more test cases that, like `CanCallInitFunction` and `CanCallFiniFunction`, test that type-erased and non type-erased code can interoperate: initialize statically, write dynamically, read statically, finalize dynamically (and other meaningful permutations).

Reply:
Testing the interoperation of type-erased and non-type-erased code is a good idea. That would help confirm that the type-erased code is actually operating on the in-memory data correctly, which is otherwise hard to do.